# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The initial stage in addressing any machine learning problem involves recognizing the data format and subsequently importing it into the chosen framework. In the context of Kaggle competitions, a significant amount of data is often available in CSV files, and that's the illustration we will employ.

We will examine a sports dataset displaying outcomes of NCAA basketball games spanning from 1985 to 2016. This dataset is stored in a CSV file, and the method we'll employ to import the file is pd.read_csv(). This function yields a dataframe variable, which is a pivotal data structure in Pandas. A dataframe is characterized as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)."

Just think of it as a table for now. 

In [2]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [3]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [4]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [6]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [8]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [9]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [10]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [11]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [12]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [13]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [14]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [16]:
#This code calculates and prints the number of unique seasons in the 'Season' column of a dataframe.

unique_seasons = len(df['Season'].unique())
print(unique_seasons)

32


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [18]:
from collections import Counter

# Count the number of wins for each team in the 'Wteam' column
win_counts = Counter(df['Wteam'])

# Find the team with the most wins by identifying the team code with the maximum count
team_with_most_wins = max(win_counts, key=win_counts.get)

# Print the team with the most wins
print("Team with the most wins:", team_with_most_wins)

Team with the most wins: 1181


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [16]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [17]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [18]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [19]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [20]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [21]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [22]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [23]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [24]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [25]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [19]:
# Sort the dataframe by the 'Season' column and assign it to a new variable
sorted_by_season = df.sort_values(by='Season')

# Sort the dataframe by the 'Wteam' column and assign it to a new variable
sorted_by_winning_team = df.sort_values(by='Wteam')

# Sort the dataframe by the 'Wscore' column and assign it to a new variable
sorted_by_winning_score = df.sort_values(by='Wscore')

# Select rows from index 100 to 200 and specific columns (Season, Wteam, Wscore)
selected_data = df.iloc[100:201, [0, 2, 3]]  # Assuming Season is column 0, Wteam is column 2, Wscore is column 3

# Print the dataframe sorted by 'Season'
print("Sorted by Season:")
print(sorted_by_season)

# Print the dataframe sorted by 'Wteam'
print("\nSorted by Winning Team:")
print(sorted_by_winning_team)

# Print the dataframe sorted by 'Wscore'
print("\nSorted by Winning Score:")
print(sorted_by_winning_score)

# Print the selected rows and columns
print("\nSelected Rows and Columns:")
print(selected_data)


Sorted by Season:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
2484      1985      99   1275      73   1132      63    H      0
2485      1985      99   1306      62   1171      59    H      0
2486      1985      99   1318      99   1145      91    H      0
2487      1985      99   1337      75   1109      56    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
141706    2016      50   1300      57   1183      52    N      0
141705    2016      50   1298      62   1202      49    H      0
141704    2016      50   1295      68   1410      50    N      0
141702    2016      50   1277      99   1324      93    N      1
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

Sorted by Winning Team:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
140642    2016      25   1101      72   1197      62    N      0
136740    2015      

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [20]:
# Find the game with the highest winning score from the sorted_by_winning_score dataframe
highest_winning_score_row = sorted_by_winning_score.iloc[0]

# Retrieve the highest winning score, season, and winning team from the original dataframe
highest_winning_score = highest_winning_score_row['Wscore']
season_of_highest_score = highest_winning_score_row['Season']
winning_team_of_highest_score = highest_winning_score_row['Wteam']

# Print the game with the highest winning score
print("Game with the Highest Winning Score:")
print("Season:", season_of_highest_score)
print("Winning Team:", winning_team_of_highest_score)
print("Winning Score:", highest_winning_score)


Game with the Highest Winning Score:
Season: 2013
Winning Team: 1264
Winning Score: 34


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [28]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [29]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [21]:
# Create a new column 'ScoreDifference' which represents the absolute difference between 'Wscore' and 'Lscore'
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Calculate the average 'ScoreDifference'
average_score_difference = df['ScoreDifference'].mean()

# Filter the DataFrame based on the condition where 'ScoreDifference' is greater than the average
filtered_data = df.loc[df['ScoreDifference'] > average_score_difference]

# Print the filtered data
print("Filtered Data:")
print(filtered_data)


Filtered Data:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...         

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [22]:
# Find the game with the highest 'ScoreDifference' from the filtered DataFrame
highest_score_difference_row = filtered_data.loc[filtered_data['ScoreDifference'].idxmax()]

# Retrieve the season, winning team, and losing team with the highest score difference
season_of_highest_score_difference = highest_score_difference_row['Season']
winning_team = highest_score_difference_row['Wteam']
losing_team = highest_score_difference_row['Lteam']

# Print the game with the highest score difference
print("Game with the Highest Score Difference:")
print("Season:", season_of_highest_score_difference)
print("Winning Team:", winning_team)
print("Losing Team:", losing_team)


Game with the Highest Score Difference:
Season: 1996
Winning Team: 1409
Losing Team: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [32]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [33]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [34]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [35]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [23]:
# Calculate the average winning score by season and store the results in a Series
average_winning_score_by_season = df.groupby('Season')['Wscore'].mean()

# Print the average winning score by season
print("Average Winning Score by Season:")
for season, average_score in average_winning_score_by_season.items():
    print(f"Season {season}: {average_score}")



Average Winning Score by Season:
Season 1985: 74.72303987155472
Season 1986: 74.81363996827915
Season 1987: 77.99386973180077
Season 1988: 79.77370417193426
Season 1989: 81.7285112707456
Season 1990: 80.84647713226205
Season 1991: 82.73393160320155
Season 1992: 79.99200387690817
Season 1993: 79.50878955298845
Season 1994: 81.1603448275862
Season 1995: 80.09614912926172
Season 1996: 78.20548277535177
Season 1997: 76.30084235860409
Season 1998: 77.22990160787137
Season 1999: 76.04358124111795
Season 2000: 76.31046691745962
Season 2001: 77.08865010073875
Season 2002: 76.9635565312843
Season 2003: 75.79506065857886
Season 2004: 74.97112229271494
Season 2005: 74.69368983957219
Season 2006: 74.58755518183729
Season 2007: 74.8294665873488
Season 2008: 74.96242494673639
Season 2009: 74.03372070870643
Season 2010: 74.65247957438723
Season 2011: 74.37418985894014
Season 2012: 73.37997334856273
Season 2013: 72.84078947368421
Season 2014: 76.10126818351361
Season 2015: 72.68079940231603
Season 201

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [24]:
# Find the maximum winning score for each team by grouping the DataFrame by winning team
max_winning_score_by_team = df.groupby('Wteam')['Wscore'].max()

# Print the maximum winning score by team
print("Maximum Winning Score by Team:")
for team, max_score in max_winning_score_by_team.items():
    print(f"Team {team}: {max_score}")


Maximum Winning Score by Team:
Team 1101: 95
Team 1102: 111
Team 1103: 109
Team 1104: 114
Team 1105: 114
Team 1106: 151
Team 1107: 108
Team 1108: 120
Team 1109: 137
Team 1110: 128
Team 1111: 111
Team 1112: 133
Team 1113: 123
Team 1114: 125
Team 1115: 106
Team 1116: 166
Team 1117: 116
Team 1118: 80
Team 1119: 104
Team 1120: 120
Team 1121: 103
Team 1122: 115
Team 1123: 115
Team 1124: 116
Team 1125: 107
Team 1126: 112
Team 1127: 94
Team 1128: 92
Team 1129: 116
Team 1130: 123
Team 1131: 110
Team 1132: 136
Team 1133: 139
Team 1134: 101
Team 1135: 115
Team 1136: 103
Team 1137: 116
Team 1138: 110
Team 1139: 144
Team 1140: 128
Team 1141: 115
Team 1142: 122
Team 1143: 121
Team 1144: 104
Team 1145: 108
Team 1146: 112
Team 1147: 124
Team 1148: 129
Team 1149: 114
Team 1150: 117
Team 1151: 125
Team 1152: 128
Team 1153: 116
Team 1154: 112
Team 1155: 120
Team 1156: 121
Team 1157: 115
Team 1158: 116
Team 1159: 111
Team 1160: 132
Team 1161: 108
Team 1162: 108
Team 1163: 130
Team 1164: 110
Team 1165: 10

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [25]:
# Group the DataFrame by both season and winning team
grouped_data = df.groupby(['Season', 'Wteam'])

# Find the team with the highest average winning score for each season
highest_avg_score_by_season = grouped_data['Wscore'].mean().groupby('Season').idxmax().apply(lambda x: x[1])

# Print the team with the highest average winning score by season
print("Team with the Highest Average Winning Score by Season:")
print(highest_avg_score_by_season)


Team with the Highest Average Winning Score by Season:
Season
1985    1328
1986    1109
1987    1380
1988    1258
1989    1258
1990    1258
1991    1380
1992    1380
1993    1380
1994    1380
1995    1206
1996    1206
1997    1254
1998    1395
1999    1317
2000    1395
2001    1395
2002    1242
2003    1395
2004    1194
2005    1311
2006    1190
2007    1440
2008    1440
2009    1377
2010    1440
2011    1449
2012    1370
2013    1377
2014    1322
2015    1322
2016    1146
Name: Wscore, dtype: int64


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [26]:
# Group the DataFrame by both season and winning team
grouped_data = df.groupby(['Season', 'Wteam'])

# Count the number of wins for each team in each season
wins_count = grouped_data.size().reset_index(name='WinsCount')

# Print the number of wins for each team in each season
print("Number of Wins for Each Team in Each Season:")
print(wins_count)


Number of Wins for Each Team in Each Season:
       Season  Wteam  WinsCount
0        1985   1102          5
1        1985   1103          9
2        1985   1104         21
3        1985   1106         10
4        1985   1108         19
...       ...    ...        ...
10167    2016   1460         20
10168    2016   1461         12
10169    2016   1462         27
10170    2016   1463         21
10171    2016   1464          9

[10172 rows x 3 columns]


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [27]:
# Group the DataFrame by season and winning team to count the number of wins
wins_count = df.groupby(['Season', 'Wteam']).size().reset_index(name='WinsCount')

# Create a mask to filter the rows with the most wins for each season
mask = wins_count.groupby('Season')['WinsCount'].transform(max) == wins_count['WinsCount']

# Filter the wins_count DataFrame to get the team with the most wins for each season
team_with_most_wins_by_season = wins_count.loc[mask, ['Season', 'Wteam']]

# Print the team with the most wins for each season
print("Team with the Most Wins for Each Season:")
for _, row in team_with_most_wins_by_season.iterrows():
    season, team = row['Season'], row['Wteam']
    print(f"Season: {season}, Team: {team}")



Team with the Most Wins for Each Season:
Season: 1985, Team: 1385
Season: 1985, Team: 1424
Season: 1986, Team: 1181
Season: 1987, Team: 1424
Season: 1988, Team: 1112
Season: 1989, Team: 1328
Season: 1990, Team: 1247
Season: 1990, Team: 1424
Season: 1991, Team: 1116
Season: 1991, Team: 1424
Season: 1992, Team: 1181
Season: 1993, Team: 1231
Season: 1993, Team: 1314
Season: 1994, Team: 1163
Season: 1994, Team: 1269
Season: 1994, Team: 1314
Season: 1995, Team: 1116
Season: 1995, Team: 1269
Season: 1996, Team: 1269
Season: 1997, Team: 1242
Season: 1998, Team: 1242
Season: 1999, Team: 1181
Season: 2000, Team: 1409
Season: 2001, Team: 1181
Season: 2002, Team: 1153
Season: 2003, Team: 1166
Season: 2003, Team: 1246
Season: 2004, Team: 1390
Season: 2005, Team: 1228
Season: 2006, Team: 1181
Season: 2006, Team: 1272
Season: 2007, Team: 1242
Season: 2007, Team: 1272
Season: 2007, Team: 1326
Season: 2008, Team: 1272
Season: 2009, Team: 1272
Season: 2010, Team: 1242
Season: 2010, Team: 1246
Season: 2

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [28]:
# Calculate the average losing score for each losing team
losing_scores = df.groupby('Lteam')['Lscore'].mean()

# Calculate the average winning score for each winning team
winning_scores = df.groupby('Wteam')['Wscore'].mean()

# Create a DataFrame to compare the average losing scores and winning scores
compare_scores = pd.DataFrame({'Avg Losing Score': losing_scores, 'Avg Winning Score': winning_scores})

# Add a column indicating if the average losing score is higher than the average winning score
compare_scores['Avg Losing Score Higher'] = compare_scores['Avg Losing Score'] > compare_scores['Avg Winning Score']

# Print the DataFrame comparing the scores
print(compare_scores)


      Avg Losing Score  Avg Winning Score  Avg Losing Score Higher
1101         60.586207          78.111111                    False
1102         59.201507          69.893204                    False
1103         64.117347          75.839768                    False
1104         64.374317          75.825944                    False
1105         61.675373          74.960894                    False
...                ...                ...                      ...
1460         63.082915          75.531469                    False
1461         64.661972          75.170082                    False
1462         68.216117          79.906021                    False
1463         60.829213          71.720102                    False
1464         65.187063          73.926056                    False

[364 rows x 3 columns]


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [29]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [30]:
import numpy as np

# Create a new column 'HighScoringGame' in dataframe df
# np.where(condition, value if true, value if false)
# Here, the condition is whether 'Wscore' > 100
# If true, assign 'Yes', otherwise assign 'No'
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

# print the DataFrame
print(df)


        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference HighScoringGame  
0                    17              No  
1                     7              No  
2                     7              No  
3              

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [32]:
# Concatenate 'Wteam' and 'Lteam' columns to count all games regardless of win/loss
all_teams = pd.concat([df['Wteam'], df['Lteam']])

# Use value_counts() method which returns a Series containing counts of unique values.
games_count = all_teams.value_counts()

# Iterate over the Series and print the results.
# games_count.index refers to the team name, and games_count.values refers to the count of games.
for team, count in zip(games_count.index, games_count.values):
    print(f"Team {team} played {count} games.")



Team 1181 played 1013 games.
Team 1314 played 1010 games.
Team 1246 played 999 games.
Team 1257 played 994 games.
Team 1242 played 993 games.
Team 1393 played 993 games.
Team 1228 played 992 games.
Team 1272 played 992 games.
Team 1437 played 983 games.
Team 1400 played 982 games.
Team 1112 played 981 games.
Team 1196 played 981 games.
Team 1116 played 980 games.
Team 1396 played 979 games.
Team 1301 played 979 games.
Team 1163 played 977 games.
Team 1281 played 977 games.
Team 1104 played 975 games.
Team 1433 played 975 games.
Team 1424 played 974 games.
Team 1210 played 972 games.
Team 1153 played 970 games.
Team 1307 played 969 games.
Team 1231 played 969 games.
Team 1268 played 969 games.
Team 1326 played 968 games.
Team 1328 played 968 games.
Team 1234 played 968 games.
Team 1277 played 966 games.
Team 1417 played 966 games.
Team 1345 played 966 games.
Team 1409 played 966 games.
Team 1140 played 965 games.
Team 1397 played 963 games.
Team 1330 played 963 games.
Team 1338 played 9

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [33]:
# Create a new column 'score_diff' which is the difference between 'Wscore' and 'Lscore'
df['score_diff'] = df['Wscore'] - df['Lscore']

# Group the DataFrame by 'Season' and find the row with maximum 'score_diff' in each group
highest_score_diff_df = df.loc[df.groupby('Season')['score_diff'].idxmax()]

# Iterate over the rows of highest_score_diff_df DataFrame and print the results
for index, row in highest_score_diff_df.iterrows():
    print(f"In Season {row['Season']}, the game with the highest score difference was between teams {row['Wteam']} and {row['Lteam']} with a difference of {row['score_diff']}.")

# Optional: Drop the 'score_diff' column if it's no longer needed
df.drop('score_diff', axis=1, inplace=True)


In Season 1985, the game with the highest score difference was between teams 1361 and 1288 with a difference of 60.
In Season 1986, the game with the highest score difference was between teams 1314 and 1264 with a difference of 84.
In Season 1987, the game with the highest score difference was between teams 1155 and 1118 with a difference of 73.
In Season 1988, the game with the highest score difference was between teams 1328 and 1147 with a difference of 68.
In Season 1989, the game with the highest score difference was between teams 1242 and 1135 with a difference of 70.
In Season 1990, the game with the highest score difference was between teams 1181 and 1217 with a difference of 76.
In Season 1991, the game with the highest score difference was between teams 1163 and 1148 with a difference of 68.
In Season 1992, the game with the highest score difference was between teams 1116 and 1126 with a difference of 82.
In Season 1993, the game with the highest score difference was between t

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [46]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [34]:
# 'melt' will reshape the DataFrame, transforming the 'Wteam' and 'Lteam' columns into rows under a new column 'team'
reshaped_df = df.melt(value_vars=['Wteam', 'Lteam'], var_name='WL', value_name='team')

# value_counts() will count the occurrences of each team in the new 'team' column
total_games_count = reshaped_df['team'].value_counts()

print("Total Number of Games Played by Each Team:")
print(total_games_count)


Total Number of Games Played by Each Team:
1181    1013
1314    1010
1246     999
1257     994
1242     993
        ... 
1230      64
1446      48
1289      48
1118      27
1327      22
Name: team, Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [35]:
# Create a new column 'ScoreDifference' with vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Find the max score difference for each season
max_score_diff_per_season = df.groupby('Season')['ScoreDifference'].max()

# Merge the dataframes on 'Season' and 'ScoreDifference' to get the details of the game with the highest score difference
highest_score_diff_game_details = df.merge(max_score_diff_per_season, how='right', 
                                           left_on=['Season', 'ScoreDifference'], 
                                           right_on=['Season', 'ScoreDifference'])

print("Game with the Highest Score Difference for Each Season:")
print(highest_score_diff_game_details)

# Optional: Drop the 'ScoreDifference' column if it's no longer needed
df.drop('ScoreDifference', axis=1, inplace=True)


Game with the Highest Score Difference for Each Season:
    Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0     1985      33   1361     128   1288      68    H      0               60   
1     1986      60   1314     129   1264      45    N      0               84   
2     1987      51   1155     112   1118      39    H      0               73   
3     1988      40   1328     152   1147      84    H      0               68   
4     1989      64   1242     115   1135      45    H      0               70   
5     1990      26   1181     130   1217      54    H      0               76   
6     1991      73   1163     115   1148      47    H      0               68   
7     1991     112   1314     118   1154      50    H      0               68   
8     1992      30   1116     128   1126      46    H      0               82   
9     1993      86   1328     146   1197      65    H      0               81   
10    1994      47   1228     121   1152      52    H

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [49]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [50]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [51]:
type(df['Wscore'])

pandas.core.series.Series

In [52]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [53]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [54]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [37]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y